src/Microsoft.OData.Client/Metadata/ODataTypeInfo.cs

```diff
@@ -0,0 +1,382 @@
+                    }
+
+                }
+
+                return _serverDefinedTypeFullName;
+            }
+        }
+
+        /// <summary>
+        /// Concurrent Dictionary cache to save ClientDefinedName for the type and serverDefinedName
+        /// </summary>
+        private ConcurrentDictionary<string, string> ServerSideNameDict { get; set; }
+
+        /// <summary>
+        /// Dictionary cache to save Propertyinfo for the type
+        /// </summary>
+        internal Dictionary<string, PropertyInfo> PropertyInfoDict
+        {
+            get
+            {
+                if (_propertyInfoDict == null)
+                {
+                    _properties = GetAllProperties();
+                }
+
+                return _propertyInfoDict;
+            }
+        }
+
+        public IEnumerable<PropertyInfo> Properties
+        {
+            get
+            {
+                if (_properties == null)
+                {
+                    _properties = GetAllProperties();
+                }
+
+                return _properties;
+            }
+        }
+
+        private IEnumerable<PropertyInfo> GetAllProperties()
+        {
+            List<PropertyInfo> properties = new List<PropertyInfo>();
+            _hasProperties = false;
+            Dictionary<string, PropertyInfo> propertyInfoDict = new Dictionary<string, PropertyInfo>();
+
+            foreach (PropertyInfo propertyInfo in type.GetPublicProperties(true /*instanceOnly*/, false))
+            {
+                //// examples where class<PropertyType>
+
+                //// the normal examples
+                //// PropertyType Property { get; set }
+                //// Nullable<PropertyType> Property { get; set; }
+
+                //// if 'Property: struct' then we would be unable set the property during construction (and have them stick)
+                //// but when its a class, we can navigate if non-null and set the nested properties
+                //// PropertyType Property { get; } where PropertyType: class
+
+                //// we do support adding elements to collections
+                //// ICollection<PropertyType> { get; /*ignored set;*/ }
+
+                //// indexed properties are not supporter because
+                //// we don't have anything to use as the index
+                //// PropertyType Property[object x] { /*ignored get;*/ /*ignored set;*/ }
+
+                //// also ignored
+                //// if PropertyType.IsPointer (like byte*)
+                //// if PropertyType.IsArray except for byte[] and char[]
+                //// if PropertyType == IntPtr or UIntPtr
+
+                //// Properties overriding abstract or virtual properties on a base type
+                //// are also ignored (because they are part of the base type declaration
+                //// and not of the derived type).
+
+                propertyInfoDict[ServerDefinedTypeName] = propertyInfo;

```

Comment:
>ServerDefinedTypeName [](http://example.com/codeflow?start=33&length=21)

shouldn't this be the server defined name for the propertyinfo, not for the type?